plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php

```diff
@@ -40,6 +63,92 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 
 		$xpath = $processor->get_xpath();
 
+		$parent_tag = $this->get_parent_tag_name( $xpath );
+		if ( 'PICTURE' === $parent_tag ) {
+			return true;
+		}
+
+		$this->pre_img_process( $processor, $context, $xpath );
+
+		$this->add_preload_link_for_img( $processor, $context, $xpath );
+
+		return true;
+	}
+
+	/**
+	 * Process a <picture> element.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_HTML_Tag_Processor  $processor HTML tag processor.
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context.
+	 *
+	 * @return bool Whether the tag should be tracked in URL Metrics.
+	 */
+	private function process_picture( OD_HTML_Tag_Processor $processor, OD_Tag_Visitor_Context $context ): bool {
+		$collected_sources = array();
+		$img_xpath         = null;
+
+		// Loop through child tags until we reach the closing </picture> tag.
+		while ( $processor->next_tag() ) {
+			$tag = $processor->get_tag();
+
+			// If we reached the closing </picture> tag, break.
+			if ( 'PICTURE' === $tag && $processor->is_tag_closer() ) {
+				break;
+			}
+
+			// Collect <source> elements.
+			if ( 'SOURCE' === $tag && ! $processor->is_tag_closer() ) {
+				$media = $processor->get_attribute( 'media' );
+				$type  = $processor->get_attribute( 'type' );
+
+				// Ensure that all <source> elements have a type attribute and no media attribute.
+				if ( null !== $media || null === $type ) {
+					return false;
```

Comment:
Problem: When this happens, the `IMG` will no longer get its necessary optimizations.

Reply:
Were the necessary optimizations happening in the `pre_img_process` function? if thats the case then as the logic of that function is now moved directly into the `process_img` function the optimizations will still happening even if this function returns here.